### ISSUE TITLE
Fix Bazel incompatible changes

### ISSUE THREAD

**Issue by laurentlb** [tweag/rules_haskell#489]
I'd like to be able to build rules_haskell with `--all_incompatible_changes`:
  `bazel test //tests:all --all_incompatible_changes`

It may not be possible to fix every issue now due to transitive dependencies, but we should try to fix the issues from this repository.

**Comment by mboes** [comment]
That would be great. But how does one go about doing this? I tried pretty hard recently, but quickly got blocked by upstream dependencies not being compliant. And since I don't know of a way to tell Bazel to ignore the upstream issues and tell me about the issues in this repository, I gave up. Is there a way to do that?

**Comment by laurentlb** [comment]
My PR is fixing multiple issues already. I'm iterating over multiple repositories, we should get to a relatively clean state soon.

I've noticed you have your own fork of skydoc. Did you send patches upstream or filed bugs? It's probably simpler if everyone is using the same documentation generator.

**Comment by mboes** [comment]
For sure. Not too happy with forking. I think the only patch in the fork is CSS styling customizations, which in skydoc last we checked was not configurable.

**Comment by Profpatsch** [comment]
What exactly is there still to be done for this repository @laurentlb? The mentioned PRs from above are already merged.

When I run `bazel test //tests:all --all_incompatible_changes`, it crashes in the builtin(!!) `bazel_tools` repository:

```
ERROR: /home/philip/.cache/bazel/_bazel_philip/cccb2d2da8416067607b8c5cff088df0/external/bazel_tools/tools/build_rules/test_rules.bzl:308:17: Traceback (most recent call last):
	File "/home/philip/.cache/bazel/_bazel_philip/cccb2d2da8416067607b8c5cff088df0/external/bazel_tools/tools/build_rules/test_rules.bzl", line 306
		rule(attrs = {"file": attr.label(mand...))}, <3 more arguments>)
	File "/home/philip/.cache/bazel/_bazel_philip/cccb2d2da8416067607b8c5cff088df0/external/bazel_tools/tools/build_rules/test_rules.bzl", line 308, in rule
		attr.label(mandatory = True, allow_files = Tr..., ...)
'single_file' is no longer supported. use allow_single_file instead. You can use --incompatible_disable_deprecated_attr_params to temporarily disable this check.
ERROR: error loading package 'tests': Extension file 'tools/build_rules/test_rules.bzl' has errors
```

Is there a way to ignore that?

**Comment by laurentlb** [comment]
I think this error disappears with Bazel 0.21, but Bazel 0.21 hits another issue (https://github.com/tweag/rules_haskell/issues/541).

Other issues:
* with `--incompatible_depset_is_not_iterable`:
```
	File "/usr/local/google/home/laurentlb/.cache/bazel/_bazel_laurentlb/ed90b0928775d29fdc34dd1dfaeb581c/external/io_tweag_rules_haskell/haskell/private/path_utils.bzl", line 236, in link_forest
		for src in srcs: ...
type 'depset' is not iterable. Use the `to_list()` method to get a list. Use --incompatible_depset_is_not_iterable=false to temporarily disable this check.
```

 * with `--incompatible_disable_tools_defaults_package`:
```
ERROR: /usr/local/google/home/laurentlb/git/rules_haskell/tests/BUILD:15:1: every rule of type _haskell_toolchain implicitly depends upon the target '//tools/defaults:crosstool', but this target could not be found because of: no such package 'tools/defaults': BUILD file not found on package path
```

 * with `--incompatible_disallow_old_style_args_add`:
```
	File "/usr/local/google/home/laurentlb/.cache/bazel/_bazel_laurentlb/ed90b0928775d29fdc34dd1dfaeb581c/external/io_tweag_rules_haskell/haskell/private/actions/compile.bzl", line 199, in _compilation_defaults
		args.add(ghc_args)
Args#add no longer accepts vectorized arguments when --incompatible_disallow_old_style_args_add is set. Please use Args#add_all or Args#add_joined.
```

 * with `--incompatible_depset_union`:
```
	File "/usr/local/google/home/laurentlb/.cache/bazel/_bazel_laurentlb/ed90b0928775d29fdc34dd1dfaeb581c/external/io_tweag_rules_haskell/haskell/private/actions/compile.bzl", line 351, in hs.toolchain.actions.run_ghc
		c.inputs + hs.extra_binaries
`+` operator on a depset is forbidden. See https://docs.bazel.build/versions/master/skylark/depsets.html for recommendations. Use --incompatible_depset_union=false to temporarily disable this check.
```

 * with `--incompatible_new_actions_api`:
```
	File "/usr/local/google/home/laurentlb/.cache/bazel/_bazel_laurentlb/ed90b0928775d29fdc34dd1dfaeb581c/external/bazel_tools/tools/build_rules/test_rules.bzl", line 34, in success_target
		ctx.new_file(ctx.genfiles_dir, exe, ".dat")
Use ctx.actions.declare_file instead of ctx.new_file. 
```

Sorry for the burden. We're now limiting the number of incompatible changes in Bazel. Also, we should get rules_haskell on our CI to detect issues early (long before a release).

**Comment by Profpatsch** [comment]
> Sorry for the burden. We're now limiting the number of incompatible changes in Bazel.

No, it’s positively not a burden. I think your team should be lauded for your principled approach to breaking API changes. Otherwise there would be an uncountable multitude of legacy problems down the line. And I certainly wouldn’t want to carry *that* burden.

**Comment by NinjaTrappeur** [comment]
Status update:

I've been working on this for the past 3 days. Before addressing the incompatible changes, we needed to address the breaking changes introduced by Bazel 0.21. This has been addressed on this branch: https://github.com/NinjaTrappeur/rules_haskell/tree/bazel21-migration

Before merging this branch to master, we need to merge a small fix on the nixpkgs bazel derivation: https://github.com/NixOS/nixpkgs/pull/53988 . Once this will be merged, we should able to pin the release commit and effectively migrate rules_haskell to Bazel 0.21.

In the meantime, I started addressing the incompatible changes on https://github.com/NinjaTrappeur/rules_haskell/tree/bazel22clean .

I have the same problem mentioned in this  comment https://github.com/tweag/rules_haskell/issues/489#issuecomment-452322945 : even with Bazel 0.21, running --incompatible-changes end up throwing errors from the builtin libraries (mostly from protobuf actually). I did not find a way to rule out the transitive dependencies errors so far...

**Comment by laurentlb** [comment]
Thanks a lot for your work.

For the protobuf issue, don't worry. If you use `--incompatible-changes`, add also `--incompatible_disable_deprecated_attr_params=false` to workaround it. I'll do something about (other rules are blocked on it too).

**Comment by mboes** [comment]
@laurentlb FYI #610 (just merged) should make rules_haskell a lot easier to integrate in a standard CI setup - if you still plan to hook up rules_haskell to Bazel's CI. We now no longer require Nix to to be installed for testing. We register binary distributions of GHC as toolchains, as well as Nixpkgs packages. The user then controls which toolchain gets selected via the `--host_platform` flag.

**Comment by NinjaTrappeur** [comment]
@laurentlb FYI, most of the incompatible changes have been fixed in our codebase. 

I had to disable a couple of checks though:

- Both `--incompatible_new_actions_api` and `--incompatible_disable_deprecated_attr_params` are failing on the protobuf library.
- `--incompatible_expand_directories` is triggering a pretty weird behavior. When activated, it creates a cycle in the dependency graph.

```
$ bazel build //... --incompatible_expand_directories 
INFO: Invocation ID: 9c4b56cc-4bb2-450d-b79a-2d12f2478332
INFO: Analysed 217 targets (286 packages loaded, 4935 targets configured).
INFO: Found 217 targets...
ERROR: rules_haskell/tools/runfiles/BUILD:7:1: in haskell_library rule //tools/runfiles:runfiles: cycle in dependency graph:
    configured target: //tools/runfiles:bin
    file: _middlemen/tools_Srunfiles_Sbin-runfiles
    action from: ActionLookupData{actionLookupKey=//tools/runfiles:bin BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=14}
    file: tools/runfiles/libHStoolsZSrunfilesZSrunfiles-ghc8.6.3.so
    action from: ActionLookupData{actionLookupKey=//tools/runfiles:runfiles BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=11}
    file: tools/runfiles/_obj/runfiles.dynamic.manifest
    action from: ActionLookupData{actionLookupKey=//tools/runfiles:runfiles BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=8}
    file: tools/runfiles/_obj/runfiles
.-> action from: ActionLookupData{actionLookupKey=//tools/runfiles:runfiles BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=2}
|   file: tools/runfiles/extra_args_runfiles_HaskellBuildLibrary
|   action from: ActionLookupData{actionLookupKey=//tools/runfiles:runfiles BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=1}
|   file: tools/runfiles/toolsZSrunfilesZSrunfiles/_iface
`-- action from: ActionLookupData{actionLookupKey=//tools/runfiles:runfiles BuildConfigurationValue.Key[6d2637c3c06d6024aad2781cf0f375e3] false, actionIndex=2}
INFO: Elapsed time: 7.512s, Critical Path: 0.54s
INFO: 23 processes: 23 linux-sandbox.
FAILED: Build did NOT complete successfully
```

You can reproduce this error with rules haskell set to `316f858fc6aebf9140cfebf6befe97b3bd55e7c0`

**Comment by Profpatsch** [comment]
I spliced out the remaining flags into a new issue, let’s mark this issue as done for the time being.
